### PR TITLE
[TFLite] Add int16x8 support for SPACE_TO_BATCH_ND and BATCH_TO_SPACE operators

### PR DESCRIPTION
**Pull request by johan-gras**
Hello,

This PR adds int16x8 support for SPACE_TO_BATCH_ND and BATCH_TO_SPACE operators.

Thanks,
Johan.

### CODE REVIEW

**Comment by gbaned** [comment]
@miaout17 Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@miaout17 Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@miaout17 Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@johan-gras Can you please sign CLA. Thank you!

**Comment by gbaned** [comment]
Hi @johan-gras Can you please sign CLA. Thank you!

**Comment by johan-gras** [comment]
@gbaned signed.

**Comment by gbaned** [comment]
Hi @johan-gras Can you please resolve conflicts? Thank you!

**Comment by abattery** [comment]
@alankelly in terms of quantized kernel implementation wise, it looks okay to me.

Could you review this pull request?

**Comment by gbaned** [comment]
Hi @johan-gras Can you please resolve conflicts? Thank you!

**Comment by Tessil** [comment]
@gbaned Thanks, I resolved the conflicts.

File: tensorflow/lite/kernels/space_to_batch_nd.cc [hunk 1]
```diff
@@ -158,6 +172,15 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
                                   op_context.output->params.zero_point);
       }
       break;
+    case kTfLiteInt16:
+      if (kernel_type == kReference) {
+        TF_LITE_SPACE_TO_BATCH_ND(reference_ops, int16_t,
+                                  op_context.output->params.zero_point);
```
[the remark this answers is not in the document]
Comment (alankelly) [reply tensorflow/lite/kernels/space_to_batch_nd.cc:178]: After discussing my my colleagues, this isn't a bug.

File: tensorflow/lite/tools/versioning/runtime_version.cc [hunk 2]
```diff
@@ -97,6 +98,7 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
            {{BuiltinOperator_BATCH_TO_SPACE_ND, 1}, "1.6.0"},
            {{BuiltinOperator_BATCH_TO_SPACE_ND, 2}, "1.14.0"},
            {{BuiltinOperator_BATCH_TO_SPACE_ND, 3}, "2.3.0"},
+           {{BuiltinOperator_BATCH_TO_SPACE_ND, 4}, "2.11.0"},
```
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/tools/versioning/runtime_version.cc]: Updated it, thanks.

File: tensorflow/lite/tools/versioning/runtime_version.cc [hunk 3]
```diff
@@ -86,6 +86,7 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
            {{BuiltinOperator_SPACE_TO_BATCH_ND, 1}, "1.6.0"},
            {{BuiltinOperator_SPACE_TO_BATCH_ND, 2}, "1.14.0"},
            {{BuiltinOperator_SPACE_TO_BATCH_ND, 3}, "2.3.0"},
+           {{BuiltinOperator_SPACE_TO_BATCH_ND, 4}, "2.11.0"},
```
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/tools/versioning/runtime_version.cc]: Updated it, thanks.

File: tensorflow/lite/kernels/space_to_batch_nd.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/kernels/space_to_batch_nd.cc:178]: The (u)int8 calls to `SpaceToBatchND` use `op_context.output->params.zero_point` as the zero-point may not be centred to 0. The int16 could replace this by a simple 0 in a way similar to the float/int32/int64 calls as we check that the int16 zero-point is null, but I think it's better to keep `op_context.output->params.zero_point` if it's alright for you as it's more explicit.

**Comment by gbaned** [comment]
Hi @alankelly Can you please review this PR ? Thank you!

**Review by alankelly — APPROVED** [review]

**Comment by gbaned** [comment]
@johan-gras Can you please resolve conflicts? Thank you!

**Review by alankelly — APPROVED** [review]

**Comment by fergushenderson** [comment]
Please add some release notes in `RELEASE.md` to the `tf.lite` section to give open source contributors details about this change when it gets included into a new release. This is particularly critical for updates to or additions of builtin operators.

**Comment by fergushenderson** [comment]
It looks like Alan Kelly already wrote an update for the RELEASE.md:

    *   Add int16x8 support for the built-in ops `space_to_batch_nd` and
        `batch_to_space_nd`

However, it looks like some tests are failing with this change, in particular 
//tensorflow/lite/testing:zip_test_batch_to_space_nd

**Comment by gbaned** [comment]
Hi @johan-gras Can you please check @fergushenderson's [comments](https://github.com/tensorflow/tensorflow/pull/52576#issuecomment-1469838640) and keep us posted ? Thank you!

**Comment by alankelly** [comment]
@Tessil The newly added tests in batch_to_space_nd.py are failing:

   {
          "dtype": [tf.float32],
          "input_shape": [[1, 3, 3, 1]],
          "block_shape": [[1, 1]],
          "crops": [[[0, 0], [0, 0]], [[1, 1], [1, 1]]],
          "constant_block_shape": [True],
          "constant_crops": [True],
          "fully_quantize": [True],
          "quant_16x8": [False, True],
      },

and 

     {
          "dtype": [tf.float32],
          "input_shape": [[1, 3, 3]],
          "block_shape": [[1]],
          "crops": [[[0, 0]], [[1, 1]]],
          "constant_block_shape": [True],
          "constant_crops": [True],
          "fully_quantize": [True],
          "quant_16x8": [False, True],
      },

 RUN      ] tests/OpsTest.RunZipTests/_batch_to_space_nd_block_shape__1_1__constant_block_shape_True_constant_crops_True_crops___0_0___0_0___dtype_tf_float32_fully_quantize_True_input_shape__1_3_3_1__quant_16x8_False
TfLiteDriver: There were errors in invocation '', validating output tensor '5':
  Tensor[BatchToSpaceND:0] index 0: got -0.994735, but expected -86.9931
  Tensor[BatchToSpaceND:0] index 1: got 1.00257, but expected 64.4628
  Tensor[BatchToSpaceND:0] index 2: got -0.994735, but expected -91.8393
  Tensor[BatchToSpaceND:0] index 3: got 1.00257, but expected 15.3672
  Tensor[BatchToSpaceND:0] index 4: got 1.00257, but expected 86.8562
  Tensor[BatchToSpaceND:0] index 5: got -0.994735, but expected -35.8177
  Tensor[BatchToSpaceND:0] index 6: got -0.994735, but expected -13.4931
  Tensor[BatchToSpaceND:0] index 7: got -0.994735, but expected -24.7565
  Tensor[BatchToSpaceND:0] index 8: got 1.00257, but expected 22.174
third_party/tensorflow/lite/testing/generated_examples_zip_test.cc:324: Failure
Value of: bug_number.empty()
  Actual: true
Expected: false
Stack trace:
0x7f80eca0fcd4: tflite::testing::OpsTest_RunZipTests_Test::TestBody() @ ??:??
0x7f7b5704bc33: testing::Test::Run() @ ??:??
0x7f7b5704cea8: testing::TestInfo::Run() @ ??:??
... Google Test internal frames ...

[  FAILED  ] tests/OpsTest.RunZipTests/_batch_to_space_nd_block_shape__1_1__constant_block_shape_True_constant_crops_True_crops___0_0___0_0___dtype_tf_float32_fully_quantize_True_input_shape__1_3_3_1__quant_16x8_False, where GetParam() = "/usr/local/google/_blaze_alankelly/06320f6a1a2cf1f03a1c833518cb9e20/execroot/google3/_tmp/cdfc8dcb58a5c01ec6b18f7e69248b68/tempfile-alankelly0.zrh.corp.google.com-161bf3-1448947-5f81ecf0bb8f0/blaze-out/k8-opt/genfiles/third_party/tensorflow/lite/testing/batch_to_space_nd_block_shape=[1,1],constant_block_shape=True,constant_crops=True,crops=[[0,0],[0,0]],dtype=tf.float32,fully_quantize=True,input_shape=[1,3,3,1],quant_16x8=False" (558 ms)


Can you please take a look at this? Thanks

**Pull request closed without merging**